pkg/keymanagementprovider/refresh/kubeRefresh.go

```diff
@@ -0,0 +1,182 @@
+		logger.Error(err, "unable to list certificate stores")
+		kr.Result = ctrl.Result{}
+		return err
+	}
+
+	if len(certificateStoreList.Items) > 0 {
+		// Note: for backwards compatibility in upgrade scenarios, Ratify will only log a warning statement.
+		logger.Warn("Certificate Store already exists. Key management provider and certificate store should not be configured together. Please migrate to key management provider and delete certificate store.")
+	}
+
+	provider, err := cutils.SpecToKeyManagementProvider(keyManagementProvider.Spec.Parameters.Raw, keyManagementProvider.Spec.Type)
+	if err != nil {
+		writeKMProviderStatus(ctx, kr, &keyManagementProvider, logger, isFetchSuccessful, err.Error(), lastFetchedTime, nil)
+		kr.Request = ctrl.Request{}
+		return err
+	}
+
+	// fetch certificates and store in map
+	certificates, certAttributes, err := provider.GetCertificates(ctx)
+	if err != nil {
+		writeKMProviderStatus(ctx, kr, &keyManagementProvider, logger, isFetchSuccessful, err.Error(), lastFetchedTime, nil)
+		kr.Request = ctrl.Request{}
+		return fmt.Errorf("error fetching certificates in KMProvider %v with %v provider, error: %w", resource, keyManagementProvider.Spec.Type, err)
+	}
+
+	// fetch keys and store in map
+	keys, keyAttributes, err := provider.GetKeys(ctx)
+	if err != nil {
+		writeKMProviderStatus(ctx, kr, &keyManagementProvider, logger, isFetchSuccessful, err.Error(), lastFetchedTime, nil)
+		kr.Request = ctrl.Request{}
+		return fmt.Errorf("error fetching keys in KMProvider %v with %v provider, error: %w", resource, keyManagementProvider.Spec.Type, err)
+	}
+	kmp.SetCertificatesInMap(resource, certificates)
+	kmp.SetKeysInMap(resource, keyManagementProvider.Spec.Type, keys)
+	// merge certificates and keys status into one
+	maps.Copy(keyAttributes, certAttributes)
+	isFetchSuccessful = true
+	emptyErrorString := ""
+	writeKMProviderStatus(ctx, kr, &keyManagementProvider, logger, isFetchSuccessful, emptyErrorString, lastFetchedTime, keyAttributes)
+
+	logger.Infof("%v certificate(s) & %v key(s) fetched for key management provider %v", len(certificates), len(keys), resource)
+
+	// returning empty result and no error to indicate we’ve successfully reconciled this object
+	// will not reconcile again unless resource is recreated
+	if !provider.IsRefreshable() {
+		kr.Request = ctrl.Request{}
+		return nil
+	}
+
+	// if interval is not set, disable refresh
+	if keyManagementProvider.Spec.Interval == "" {
+		logger.Infof("KeyManagementProvider %v is refreshable but interval is not set", resource)
+		kr.Result = ctrl.Result{}
+		return nil
+	}
+	// resource is refreshable, requeue after interval
+	intervalDuration, err := time.ParseDuration(keyManagementProvider.Spec.Interval)
+	if err != nil {
+		logger.Error(err, "unable to parse interval duration")
+		kr.Result = ctrl.Result{}
+		return err
+	}
+
+	logger.Info("Reconciled KeyManagementProvider", "intervalDuration", intervalDuration)
```

Comment:
@akashsinghal what do you think about logging each time it refreshes? If the interval is low, this is going to generate A LOT of stuff. These changes are visible in the resource itself with LastFetchedTime and the version of the certificate.

Reply:
I'm thinking a short message for refresh is ok in our logs at INFO level. In general I'd assume the interval is likely going to be in the order of minutes. I'll defer to your judgement if you'd like to add it or not.